crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -1141,6 +1045,390 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
+                    // the first one, nothing to do.
+                }
+
+                prev_item = Some((item, *ts));
+            }
+
+            if let Some(event) = item.as_event() {
+                // It's an event.
+                let ts = event.timestamp();
+                let event_date = timestamp_to_date(ts);
+
+                if let Some((prev_item, prev_ts)) = prev_item {
+                    if prev_item.is_day_divider() {
+                        // The event is preceded by a day divider.
+                        if timestamp_to_date(prev_ts) != event_date {
+                            // The day divider is wrong. Should we replace it with the correct
+                            // value, or remove it entirely?
+                            if last_event_ts.map_or(
+                                false,
+                                |last_event_ts: MilliSecondsSinceUnixEpoch| {
+                                    timestamp_to_date(last_event_ts) == event_date
+                                },
+                            ) {
+                                // There's a previous event with the same date: remove the divider.
+                                trace!("removed day divider @ {i} between two events that have the same date");
+                                ops.push(DayDividerOperation::Remove(i - 1));
+                            } else {
+                                // There's no previous event or there's one with a different date:
+                                // replace the current divider.
+                                trace!("replacing day divider @ {i} with new timestamp from event");
+                                ops.push(DayDividerOperation::Replace(i - 1, ts));
+                            }
+                        }
+                    } else if let Some(prev_event) = prev_item.as_event() {
+                        // The event is preceded by another event. If they're not the same date,
+                        // insert a date divider.
+                        let prev_ts = prev_event.timestamp();
+                        if timestamp_to_date(prev_ts) != timestamp_to_date(ts) {
+                            trace!("inserting day divider @ {} between two events with different dates", i);
+                            ops.push(DayDividerOperation::Insert(i, ts));
+                        }
+                    }
+                } else {
+                    // The event was the first item, so there wasn't any day divider before it:
+                    // insert one.
+                    trace!("inserting the first day divider @ {}", i);
+                    ops.push(DayDividerOperation::Insert(i, ts));
+                }
+
+                prev_item = Some((item, ts));
+                last_event_ts = Some(ts);
+            }
+        }
+
+        //#[cfg(debug)]
+        // TODO: in non-debug mode, set to `None` instead.
```

Comment:
TODO

Reply:
Not sure what to do here:

- the previous code would clone the items if in debug, and not in non-debug; that's nice for tests, but would make rageshakes in apps useless
- cloning all the time might have a small cost: it's only a vec of `Arc`, but this would happen every time we add new events to a timeline
- the midway could be to add some special mode to the timeline/SDK that would enable heavyweight logs, and if enabled, we'd clone here

Reply:
…or just do it if we're at the trace log level only \o/